src/UserSpaceInstrumentation/Trampoline.cpp

```diff
@@ -0,0 +1,155 @@
+
+using absl::numbers_internal::safe_strtou64_base;
+using orbit_base::ReadFileToString;
+
+bool DoAddressRangesOverlap(const AddressRange& a, const AddressRange& b) {
+  return !(b.end <= a.start || b.start >= a.end);
+}
+
+std::optional<size_t> LowestIntersectingAddressRange(const std::vector<AddressRange>& ranges_sorted,
+                                                     const AddressRange& range) {
+  for (size_t i = 0; i < ranges_sorted.size(); i++) {
+    if (DoAddressRangesOverlap(ranges_sorted[i], range)) {
+      return i;
+    }
+  }
+  return std::nullopt;
+}
+
+std::optional<size_t> HighestIntersectingAddressRange(
+    const std::vector<AddressRange>& ranges_sorted, const AddressRange& range) {
+  for (int i = ranges_sorted.size() - 1; i >= 0; i--) {
+    if (DoAddressRangesOverlap(ranges_sorted[i], range)) {
+      return i;
+    }
+  }
+  return std::nullopt;
+}
+
+ErrorMessageOr<std::vector<AddressRange>> GetTakenAddressRanges(pid_t pid) {
+  std::vector<AddressRange> result;
+  OUTCOME_TRY(mmap_min_addr, ReadFileToString("/proc/sys/vm/mmap_min_addr"));
+  uint64_t mmap_min_addr_as_uint64 = 0;
+  if (!absl::SimpleAtoi(mmap_min_addr, &mmap_min_addr_as_uint64)) {
+    return ErrorMessage("Failed to read/proc/sys/vm/mmap_min_addr");
+  }
+  result.emplace_back(0, mmap_min_addr_as_uint64);
+
+  OUTCOME_TRY(maps, ReadFileToString(absl::StrFormat("/proc/%d/maps", pid)));
+  std::vector<std::string> lines = absl::StrSplit(maps, '\n', absl::SkipEmpty());
+  for (const auto& line : lines) {
+    std::vector<std::string> tokens = absl::StrSplit(line, ' ', absl::SkipEmpty());
+    if (tokens.size() < 1) {
+      continue;
+    }
+    std::vector<std::string> addresses = absl::StrSplit(tokens[0], '-');
+    if (addresses.size() != 2) {
+      continue;
+    }
+    uint64_t address_begin = 0;
+    uint64_t address_end = 0;
+    if (!safe_strtou64_base(addresses[0], &address_begin, 16) ||
+        !safe_strtou64_base(addresses[1], &address_end, 16)) {
+      continue;
+    }
+    CHECK(address_begin < address_end);
+    // Join with previous segment ...
+    if (result.back().end == address_begin) {
+      result.back().end = address_end;
+      continue;
+    }
+    // ... or append as new segment.
+    result.emplace_back(address_begin, address_end);
+  }
+  return result;
+}
+
+ErrorMessageOr<AddressRange> FindAddressRangeForTrampoline(
+    const std::vector<AddressRange>& taken_ranges, const AddressRange& code_range, uint64_t size) {
+  constexpr uint64_t kMax32BitOffset = 0x7fffffff;
+  const uint64_t page_size = sysconf(_SC_PAGE_SIZE);
+
+  // Try to fit an interval of length `size` below `code_range`.
+  auto optional_range_index = LowestIntersectingAddressRange(taken_ranges, code_range);
+  if (!optional_range_index) {
+    return ErrorMessage(absl::StrFormat("code_range %#x-%#x is not in taken_ranges.",
+                                        code_range.start, code_range.end));
+  }
+  while (optional_range_index.value() > 0) {
```

Comment:
this implies we know here that the very first segment is always starts with 0 address. Maybe add a comment about that?

Reply:
and probably a check since things could go wrong if someone passes list of addresses without 0-min_vaddr

Reply:
Hmm, the method is *really* specific to the output of GetUnavailableAddressRanges. Strictly speaking I could also add a check for non overlapping regions. The only reason they are not tied together (either calling GetUnavailableAddressRanges from within FindAddressRangeForTrampoline or at least putting GetUnavailableAddressRanges  into an unnnamed namespace) is that I'd like to have them available for testing.

Added a check and comment. done.